server.c

```diff
@@ -1292,6 +1297,96 @@ client_close(struct client *c)
 	client_close_ev(c->fd, 0, c);
 }
 
+static ssize_t
+read_cb(struct tls *ctx, void *buf, size_t buflen, void *cb_arg)
+{
+	struct client *c = cb_arg;
+
+	if (!c->should_buffer) {
+		// no buffer to cache into, read into libtls buffer
+		ssize_t ret = read(c->fd, buf, buflen);
+		if (-1 == ret && errno == EWOULDBLOCK)
+			ret = TLS_WANT_POLLIN;
+		
+		return ret;
+	}
+
+	if (c->buf.has_tail) {
+		// we have leftover data from a previous call to read_cb
+		size_t left = BUFLAYER_MAX - c->buf.read_pos;
+		size_t copy_len = MINIMUM(buflen, left);
+		memcpy(buf, c->buf.data + c->buf.read_pos, copy_len);
+
+		c->buf.read_pos += copy_len;
+
+		if (left == copy_len) {
+			// memset(buflayer, 0, BUFLAYER_MAX);
+			c->should_buffer = 0;
+			c->buf.has_tail = 0;
+		}
+		
+		return copy_len;
+	}
+
+	// buffer layer exists, we expect proxy protocol
+	errno = 0;
+	ssize_t n_read = read(
+		c->fd, 
+		c->buf.data + c->buf.len, 
+		BUFLAYER_MAX - c->buf.len
+	);
+	if (-1 == n_read && errno == EWOULDBLOCK)
+		return TLS_WANT_POLLIN;
+
+	c->buf.len += n_read;
+
+	struct proxy_protocol_v1 pp1 = {0};
+	size_t consumed = 0;
+	
+	int parse_status = proxy_proto_v1_parse(&pp1, c->buf.data, c->buf.len, &consumed);
+	if (PROXY_PROTO_PARSE_SUCCESS != parse_status) {
+		close(c->fd);
+		return 0;
+	}
+
+	switch (pp1.proto) {
+		case PROTO_V4: inet_ntop(AF_INET, &pp1.srcaddr.v4, c->rhost, NI_MAXHOST); break;
+		case PROTO_V6: inet_ntop(AF_INET6, &pp1.srcaddr.v6, c->rhost, NI_MAXHOST); break;
+		case PROTO_UNKNOWN: strncpy(c->rhost, "UNKNOWN", NI_MAXHOST); break;
```

Comment:
strncpy(3) is unfortunately one of those strange functions it's hard to use and are only seldomly useful.  TLDR; in this case it should be `strlcpy(c->host, "UNKNOWN", sizeof(c->host))`.

(str**N**cpy takes the maximum length of the second argument, not the first.  in this case there's plenty of space so in practice it will never fail.  Still, it's the wrong function)

Reply:
`c->host` or `c->rhost`?

Reply:
I meant `c->rhost`, sorry.  `host` is the requested host from the user, `rhost` the remote host from where the client is connecting.  Actually the naming could be better, it's a bit confusing :)

Reply:
also, regarding the different APIs: there's the very old `strcpy` and `strcat` which assume infinite storage and so they're usually a bad sign when you seem them.  `strncpy` and `strncat` are weird because they assume the string you're copying to has infinite space but allows you to specify the length of the string you're copying from.  There could be some niche use for these, but they're usually a mistake too when you see someone using it.

`strlcpy` and `strlcat` IMHO are saner since they allow you to specify the length of the storage you're copying to.